crates/debugger/src/lib.rs

```diff
@@ -0,0 +1,636 @@
+//! Wasmtime debugger functionality.
+//!
+//! This crate builds on top of the core Wasmtime crate's
+//! guest-debugger APIs to present an environment where a debugger
+//! runs as a "co-running process" and sees the debugee as a a
+//! provider of a stream of events, on which actions can be taken
+//! between each event.
+//!
+//! In the future, this crate will also provide a WIT-level API and
+//! world in which to run debugger components.
+
+use std::{any::Any, sync::Arc};
+use tokio::{
+    sync::{Mutex, mpsc},
+    task::JoinHandle,
+};
+use wasmtime::{
+    AsContextMut, DebugEvent, DebugHandler, ExnRef, OwnedRooted, Result, Store, StoreContextMut,
+    Trap,
+};
+
+/// A `Debugger` wraps up state associated with debugging the code
+/// running in a single `Store`.
+///
+/// It acts as a Future combinator, wrapping an inner async body that
+/// performs some actions on a store. Those actions are subject to the
+/// debugger, and debugger events will be raised as appropriate. From
+/// the "outside" of this combinator, it is always in one of two
+/// states: running or stopped. When stopped, it acts as a
+/// `StoreContextMut` and can allow examining the stopped execution's
+/// state. One runs until the next event suspends execution by
+/// invoking `Debugger::run`.
+///
+/// Note that because of limitations in Wasmtime's future cancelation
+/// handling, all events must be consumed until the inner body
+/// completes and `Debugger::is_complete` returns
+/// true. `Debugger::finish` continues execution ignoring all further
+/// events to allow clean completion if needed.
+pub struct Debugger<T: Send + 'static> {
+    /// The inner task that this debugger wraps.
+    inner: Option<JoinHandle<Result<Store<T>>>>,
+    /// State: either a task handle or the store when passed out of
+    /// the complete task.
+    state: DebuggerState,
+    in_tx: mpsc::Sender<Command<T>>,
+    out_rx: mpsc::Receiver<Response>,
+}
+
+/// State machine from the perspective of the outer logic.
+///
+/// The intermediate states here, and the separation of these states
+/// from the `JoinHandle` above, are what allow us to implement a
+/// cancel-safe version of `Debugger::run` below.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum DebuggerState {
+    /// Inner body is running in an async task and not in a debugger
+    /// callback. Outer logic is waiting for a `Response::Stopped` or
+    /// `Response::Complete`.
+    Running,
+    /// Inner body is running in an async task and at a debugger
+    /// callback (or in the initial trampoline waiting for the first
+    /// `Continue`). `Response::Stopped` has been received. Outer
+    /// logic has not sent any commands.
+    Stopped,
```

Comment:
Maybe name this `Paused`, because the name is a bit ambiguous with `Complete` otherwise (a program stops when it is finished, does it not?)